### PR TITLE
S4583 FP fix with EndInvoke as MemberAccess

### LINKED ISSUE #4255: S4583: FP with BeginInvoke / EndInvoke - Endinvoke as delegate
### Description

If Action.Endinvoke is used directly as an argument of BeginInvoke, an issue is raised. (S4583)

### Repro steps

```
Action a = () => { };
a.BeginInvoke(a.EndInvoke, null);
```

### Expected behavior

An issue is not raised.

### Actual behavior

An issue is raised.

### Known workarounds

```
Action a = () => { };
a.BeginInvoke(ar => a.EndInvoke(ar), null);
```

### Related information

* C#/VB.NET Plugins version : 8.20

### PR DESCRIPTION
**Pull request by csaba-sagi-sonarsource**
Fixes #4255

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/BeginInvokePairedWithEndInvoke.cs [hunk 1]
```diff
@@ -60,14 +60,30 @@ public sealed class BeginInvokePairedWithEndInvoke : BeginInvokePairedWithEndInv
...
             new InvocationWalker(context).SafeVisit(context.Root);
 
+        protected override bool IsInvalidCallback(SyntaxNode callbackArg, SemanticModel semanticModel)
+        {
+            if (FindCallback(callbackArg, semanticModel) is { } callback)
+            {
+                if (Language.Syntax.Kind(callback) == SyntaxKind.SimpleMemberAccessExpression)
+                {
+                    return !(semanticModel.GetSymbolInfo(callback).Symbol is IMethodSymbol methodSymbol
+                             && methodSymbol.Name == EndInvoke);
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/BeginInvokePairedWithEndInvoke.cs:70]: Can be simplified a bit:
```
return !(semanticModel.GetSymbolInfo(callback).Symbol is IMethodSymbol {Name: EndInvoke});
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/BeginInvokePairedWithEndInvoke.cs [hunk 2]
```diff
@@ -60,14 +60,30 @@ public sealed class BeginInvokePairedWithEndInvoke : BeginInvokePairedWithEndInv
         protected override void VisitInvocation(EndInvokeContext context) =>
             new InvocationWalker(context).SafeVisit(context.Root);
 
+        protected override bool IsInvalidCallback(SyntaxNode callbackArg, SemanticModel semanticModel)
+        {
+            if (FindCallback(callbackArg, semanticModel) is { } callback)
+            {
+                if (Language.Syntax.Kind(callback) == SyntaxKind.SimpleMemberAccessExpression)
+                {
+                    return !(semanticModel.GetSymbolInfo(callback).Symbol is IMethodSymbol methodSymbol
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/BeginInvokePairedWithEndInvoke.cs:69]: I think we can first check the name at the syntax level and only later get the symbol, in order to improve performance.

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.CSharp/Rules/BeginInvokePairedWithEndInvoke.cs [hunk 3]
```diff
@@ -60,14 +60,26 @@ public sealed class BeginInvokePairedWithEndInvoke : BeginInvokePairedWithEndInv
         protected override void VisitInvocation(EndInvokeContext context) =>
             new InvocationWalker(context).SafeVisit(context.Root);
 
+        protected override bool IsInvalidCallback(SyntaxNode callbackArg, SemanticModel semanticModel)
+        {
+            if (FindCallback(callbackArg, semanticModel) is { } callback)
+            {
+                return callback is MemberAccessExpressionSyntax { } memberAccess && memberAccess.Name.ToString().Equals(EndInvoke)
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/BeginInvokePairedWithEndInvoke.cs:67]: pedantic: `{ }` is redundant (empty property pattern clause)

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!

**Comment by costin-zaharia-sonarsource** [comment]
Isn't this reproducible also for VB?

**Comment by csaba-sagi-sonarsource** [comment]
> Isn't this reproducible also for VB?

No, VB has no such syntax, so it is not reproducible for VB.

**Pull request merged**